compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -3766,69 +3806,100 @@ abstract class GenJSCode extends plugins.PluginComponent
-                  fromAny(callTrg,
-                      reflBoxClassPatched.primaryConstructor.tpe.params.head.tpe)
-                val reflBox = genNew(reflBoxClassPatched,
-                    reflBoxClassPatched.primaryConstructor, List(castCallTrg))
-                genApplyMethod(
-                    reflBox,
-                    proxyIdent,
-                    arguments,
-                    jstpe.AnyType)
-              }
+              // Otherwise, we need to call it on RuntimeString
+              val (rtModuleClass, methodIdent) =
+                encodeRTStringMethodSym(methodInStringClass)
+              val retTpe = methodInStringClass.tpe.resultType
+              val castCallTrg = fromAny(callTrg, StringClass.toTypeConstructor)
+              boxIfNeeded(
+                  genApplyMethod(
+                      genLoadModule(rtModuleClass),
+                      methodIdent,
+                      castCallTrg :: arguments,
+                      toIRType(retTpe)),
+                  retTpe)
             }
           }, { // else
             callStatement
           })(jstpe.AnyType)
         }
+
+        /* For primitive types, we need to handle two cases. The method could
+         * either be defined in the boxed class of the primitive type, or it
+         * could be defined in the primitive class itself. In the former case,
+         * we generate a direct call to the actual (hijacked) method. In the
+         * latter case, we directly generate the primitive operation.
+         */
+
+        def addCallForPrimitive(primitiveClass: Symbol): Boolean = {
+          val boxedClass = definitions.boxedClass(primitiveClass)
+          val boxedTpe = boxedClass.tpe
+          val methodInBoxedClass = matchingSymIn(boxedClass)
+          if (methodInBoxedClass != NoSymbol && methodInBoxedClass.isPublic) {
+            callStatement = js.If(genIsInstanceOf(callTrg, boxedTpe), {
+              boxIfNeeded(
+                  genApplyMethod(
+                      genAsInstanceOf(callTrg, boxedTpe),
+                      methodInBoxedClass,
+                      arguments),
+                  methodInBoxedClass.tpe.resultType)
+            }, { // else
+              callStatement
+            })(jstpe.AnyType)
+            true
+          } else {
+            val methodInPrimClass = matchingSymIn(primitiveClass)
+            if (methodInPrimClass != NoSymbol && methodInPrimClass.isPublic) {
+              def isIntOrLongKind(kind: TypeKind) = kind match {
+                case CharKind     => false
+                case _:INT | LONG => true
+                case _            => false
+              }
+              val ignoreBecauseItMustBeAnInt = {
+                primitiveClass == DoubleClass &&
+                toTypeKind(methodInPrimClass.tpe.resultType) == DoubleKind &&
+                isIntOrLongKind(toTypeKind(sym.tpe.resultType))
+              }
+              if (ignoreBecauseItMustBeAnInt) {
```

Comment:
In that case, `toTypeKind(methodInPrimClass.tpe.resultType)` is `IntKind`, so the second line of the condition is false.